Setup rom decompression and compression (#1614)
zeldaret/oot: README.md

It builds the following ROM:

> * oot_uncompressed-gc-eu-mq-dbg.z64 `md5: f0b7f35375f9cc8ca1b2d59d78e35405`

Reviewer: It seems a bit strange to me that final output ROM, `oot-gc-eu-mq-dbg.z64`, does not match the debug ROM. What about a `COMPRESS` makefile variable, where if COMPRESS=0 then the uncompressed ROM is directly copied to the output ROM? It would be set to 1 for retail versions and 0 for debug, but modders could override it if they want their mod to be compressed.

Author: Personally I think running `make compress` and getting an uncompressed rom on some versions but not on others is bad.
I would like to hear more people's opinions about this

Reviewer 2: Im not sure the best approach to solve this, but I do think itd be a bit weird if `oot-gc-eu-mq-dbg.z64` is not the name for the definitive matching debug rom

Reviewer 2: I tbh am not sure where I stand on this. 
Theres two lenses to look at this through: OoT historical documentation/decomp mindset, and a general use (modding) codebase.

Logically it makes sense to label the uncompressed rom as uncompressed, and I do recognize that, but from an OoT "historian" viewpoint, it makes sense for the "not standard" format to have the "non standard" name.

Reviewer 2: is `COMPARE` at the top supposed to be `COMPRESS`?
and  "oot-$(VERSION)-uncompressed.z64 is the "compressed ROM"", I assume you mean is the uncompressed rom?

So if I am following, for gc-eu-mq-dbg, we are going to get both a "uncompressed rom" and an "output rom", and both are going to be the exact same in that specific case?

Reviewer: Yes, apologies for the sloppiness.

Yeah, if you do not want compression, the output rom is an exact copy of the uncompressed rom

Reviewer 2: ok, I think this is the best sounding compromise so far

Reviewer 3: > Logically it makes sense to label the uncompressed rom as uncompressed

I don't agree. Generating the rom in it's decompressed form should be considered standard output for compilation, so it shouldn't need "uncompressed" tacked on to the end. On the other hand, compressed roms should be labeled as being compressed, since it makes it more clear that a transformation on the original data has occurred. 

Arguably, we may also want to specify the type of compression that has been applied to the rom, since not all versions use the same compression algorithm.

Reviewer 2: I keep flip flopping back and forth here, but I guess I do agree that "uncompressed" should probably be unlabeled, and compressed which has data altered should be labeled as such.
I do not agree on including the compression algorithm in the name tho, I dont find it necessary

Author: My way of seeing it is since the final result of the build process will produce a rom that matches the contents of the cartridge then that one should be favored and be unlabelled.
Another alternative would be to label both: `oot-compressed.z64` and `oot-uncompressed.z64`

Reviewer 2: Yeah I mentioned above that the 2 main ways of looking at it above is "oot rom historian" or "result from a generic build pipeline". And while I understand the historian viewpoint, I think just naming them after how it objectively is (the "game" and the "compressed game"), is much more consistent and better long term.

we can try to document somewhere which ones are originally not compressed and compressed

Reviewer 2: I do recognize the compromise option of labeling both, but do think mzx has a point here that labeling something as "uncompressed" seems a little redundant. It is just the output of make which builds the game. It get compressed after the fact to create a new variation of that original output. So I think "uncompressed" being unlabeled does make alot of sense here

* build-essential
* binutils-mips-linux-gnu
* python3
* pip3

> * venv

Reviewer: * python3-pip
* python3-venv

cd oot
```

#### 3. Install python dependencies

The build process has a few python packages required that are located in `requirements.txt`.

It is recommend to setup a virtual environment for python to localize all dependencies. To create a virtual environment:

```bash
python3 -m venv .venv
```

To start using the virtual environment on your current terminal run:

```bash
. .venv/bin/activate
```

Take in mind for each new terminal you'll need to **active** the Python virtual environment again. There's no need to create the virtual environment again.

Now you can install the Python dependencies, to do so run:

```bash

> python3 -m pip install -U -r requirements.txt

Reviewer: Why not `pip install -r requirements.txt`? That's what basically every other Python project uses so it's ingrained in my muscle memory

Author: Using `pip` instead of `python3 -m pip` can be problematic if you have more than one version of Python installed, which can happen if either you install it manually or your distro decides to push a newer version.
It is better practice to do this, but if anybody wants to just type `pip` then it should work in most cases.

As a safe-ish option you can slap `alias pip="python3 -m pip"` on your `~/.bashrc` file and forget about this issue.

Reviewer 4: Yeah I also wanted to suggest `pip` over `python3 -m pip` but couldn't find solid reference on it. Everything I found even used `python3 -m pip`, including the "official" pip tutorial

Reviewer: `pip` here refers to the command in the virtualenv, not the distro one

Reviewer 4: > `pip` here refers to the command in the virtualenv, not the distro one

Yeah I haven't found reference on that

I do use it typically but since I can't find it documented I wouldn't advocate document using it

Reviewer: The official docs at https://docs.python.org/3/library/venv.html say

> Changed in version 3.4: Installs pip by default, added the `--without-pip` and `--copies` options

and later it suggests using `pip install -r requirements.txt` (though oddly, only in a warning)

Reviewer: bump

Reviewer 5: https://poonam-jangid01.medium.com/to-clarify-when-you-use-it-you-are-invoking-the-pip-module-as-a-script-within-the-python-13170aa3f59f seems to be a good write up about the differences between them both. Tbh I don't really have a preference either way and not sure it really matters which we document (though `python3 -m pip` seems the "safer" option.)

Reviewer: That doesn't seem to be a particularly high-quality source. What does "multiple virtual environments" even mean? If you could have multiple virtual environments, wouldn't `python` be just as ambiguous as `pip`?

Reviewer 5: > What does "multiple virtual environments" even mean?

You can create as many virtual environments as you would like (different python versions, different packages based on projects), so it is very possible people have multiple virtual environments "installed" though only 1 can be active.
As stated in the article `python` will always point to the current activated python interpreter, so you can be assured which python is being used. On the other hand if you do use `pip` it will use the first pip it finds in the system path. When activating a virtual environment it puts that venv's bin folder first in the path so if it has pip, it would be fine since that pip should be found first. However if it doesn't, it would mess with whatever pip it does find on your system path which might cause issues.
New venv's seem to have pip installed by default now, so again either should work for our use case.

```

#### 4. Prepare a base ROM

Copy over your copy of the Master Quest (Debug) ROM inside the root of this new project directory.

> Rename the file to "baserom.gc-eu-mq-dbg.z64", "baserom_original.n64" or "baserom_original.v64", depending on the original extension.

Reviewer: Is this still accurate?

cd oot
```

#### 3. Install python dependencies

The build process has a few python packages required that are located in `requirements.txt`.

> It is recommend to setup a virtual environment for python to localize all dependencies. To create a virtual environment:

Reviewer 4: ```suggestion
It is recommended to set up a virtual environment for python to contain all dependencies. To create a virtual environment:
```

```bash
python3 -m venv .venv
```

> To start using the virtual environment on your current terminal run:

Reviewer 4: ```suggestion
To start using the virtual environment in your current terminal run:
```

```bash
. .venv/bin/activate
```

> Take in mind for each new terminal you'll need to **active** the Python virtual environment again. There's no need to create the virtual environment again.
> 

Reviewer 2: ```suggestion
Keep in mind for each new terminal session you'll need to **activate** the Python virtual environment again. There's no need to create the virtual environment again.

```

> Keep in mind for each new terminal session you'll need to **activate** the Python virtual environment again. There's no need to create the virtual environment again.

Reviewer 4: ```suggestion
Keep in mind that for each new terminal session, you will need to **activate** the Python virtual environment again. That is, run the above `. .venv/bin/activate` command.
```

Now you can install the Python dependencies, to do so run:

```bash
python3 -m pip install -U -r requirements.txt
```

#### 4. Prepare a base ROM

> Copy over your copy of the Master Quest (Debug) ROM inside the `baseroms/gc-eu-mq-dbg/` path of this new project directory.

Reviewer 4: while we're at it, may as well
```suggestion
Place a copy of the Master Quest (Debug) ROM inside the `baseroms/gc-eu-mq-dbg/` folder.
```

Rename the file to `baserom.z64`, `baserom.n64` or `baserom.v64`, depending on the original extension.

#### 5. Setup the ROM and build process

Setup and extract everything from your ROM with the following command:

```bash
make setup
```

> This will generate a new ROM called "baseroms/gc-eu-mq-dbg/baserom-uncompressed.z64" that will have the overdump removed and the header patched.

Reviewer 4: ```suggestion
This will generate a new ROM "baseroms/gc-eu-mq-dbg/baserom-uncompressed.z64" that will have the overdump removed and the header patched.
```
